consensus/src/round_manager_test.rs

```diff
@@ -252,40 +275,300 @@ impl NodeSetup {
+                if let Some(request) = maybe_request {
+                    info!(
+                        "RPC request received: {:?} on {:?}",
+                        request,
+                        node.identity_desc()
+                    );
+                    node.block_store
+                        .process_block_retrieval(request)
+                        .await
+                        .unwrap();
+                } else {
+                    tokio::time::sleep(Duration::from_millis(50)).await;
+                }
+            }
+            node
+        }));
+    }
+    ReplyingRPCHandle { handles, done }
+}
+
+struct ReplyingRPCHandle {
+    handles: Vec<JoinHandle<NodeSetup>>,
+    done: Arc<AtomicBool>,
+}
+
+impl ReplyingRPCHandle {
+    async fn join(self) -> Vec<NodeSetup> {
+        self.done.store(true, Ordering::Relaxed);
+        let mut result = Vec::new();
+        for handle in self.handles.into_iter() {
+            result.push(handle.await.unwrap());
+        }
+        info!(
+            "joined nodes in order: {:?}",
+            result.iter().map(|v| v.id).collect::<Vec<_>>()
+        );
+        result
+    }
+}
+
+async fn process_and_vote_on_proposal(
+    nodes: &mut [NodeSetup],
+    next_proposer: usize,
+    down_nodes: &[usize],
+    process_votes: bool,
+    apply_commit_prev_proposer: Option<usize>,
+    apply_commit_on_votes: bool,
+    expected_round: u64,
+    expected_qc_ordered_round: u64,
+    expected_qc_committed_round: u64,
+) {
+    info!(
+        "Called {} with current {} and apply commit prev {:?}",
+        expected_round, next_proposer, apply_commit_prev_proposer
+    );
+    let mut num_votes = 0;
+
+    for node in nodes.iter_mut() {
+        if down_nodes.contains(&node.id) {
+            // Drop the proposal on down nodes
+            info!("next_proposal on {}", node.identity_desc());
+            node.next_proposal().await;
+            info!("done with next_proposal on {}", node.identity_desc());
+        } else {
+            // Proccess proposal on other nodes
+            let proposal_msg = node.next_proposal().await;
+
+            assert_eq!(proposal_msg.proposal().round(), expected_round);
+            assert_eq!(
+                proposal_msg.sync_info().highest_ordered_round(),
+                expected_qc_ordered_round
+            );
+            assert_eq!(
+                proposal_msg.sync_info().highest_commit_round(),
+                expected_qc_committed_round
+            );
+
+            info!("Process proposal on {}", node.identity_desc());
+            node.round_manager
+                .process_proposal_msg(proposal_msg)
+                .await
+                .unwrap();
+            info!("Finish process proposal on {}", node.identity_desc());
+            num_votes += 1;
+
+            if let Some(prev_proposer) = apply_commit_prev_proposer {
+                if prev_proposer != node.id && expected_round > 2 {
```

Comment:
why prev_proposer != node.id?

Reply:
proposer commits a round on line 559, right after it receives the votes